arborist/engine_api.go

```diff
@@ -0,0 +1,269 @@
+		JSON:    errorJSON.marshal(),
+	}
+}
+
+// roleNotExists returns an operation indicating the requested role to operate
+// on does not exist (404).
+func roleNotExists(roleID string) ArboristOperation {
+	msg := fmt.Sprintf("no role exists with ID: %s", roleID)
+	code := http.StatusNotFound
+	errorJSON := ErrorJSON{errorInformation{
+		Message: msg,
+		Code:    code,
+	}}
+	return ArboristOperation{
+		Success: false,
+		Status:  code,
+		JSON:    errorJSON.marshal(),
+	}
+}
+
+// HandleResponseWriter takes a `ResponseWriter` and write the correct headers,
+// status, and response from the results of the operation.
+func (operation ArboristOperation) HandleResponseWriter(w http.ResponseWriter) {
+	w.WriteHeader(operation.Status)
+	w.Header().Set("Content-Type", "application/json")
+	w.Write(operation.JSON)
+}
+
+// ResourcesList stores a list of resources for a response.
+type ResourcesList struct {
+	Resources []string `json:"resources"`
+}
+
+// Wrapper struct to store an error to convert to JSON. (This is just so the
+// JSON can look like `{"error": {...}}`.)
+type ErrorJSON struct {
+	Error errorInformation `json:"error"`
+}
+
+// marshal dumps the error into JSON and assume that this will work correctly,
+// ignoring the possible error from the JSON marshalling (so assuming we have
+// defined the error JSON structs correctly).
+func (errorJSON ErrorJSON) marshal() []byte {
+	bytes, _ := marshal(errorJSON)
+	return bytes
+}
+
+// errorInformation wraps info for error message fields in a JSON response.
+type errorInformation struct {
+	Message string `json:"message"`
+	Code    int    `json:"code"`
+}
+
+// ListResources returns the list of just the names of all the roles that exist
+// in the engine.
+func (engine *AuthEngine) ListResources() ArboristOperation {
+	resources := ResourcesList{
+		Resources: engine.ListResourceNames(),
+	}
+	bytes, err := marshal(resources)
+	if err != nil {
+		return failedMarshal(err)
+	}
+
+	return ArboristOperation{
+		Success: true,
+		Status:  http.StatusOK,
+		JSON:    bytes,
+	}
+}
+
+// CreateRole, given some bytes as input which should be a JSON describing the
+// role to create, adds the new role to the engine.
+func (engine *AuthEngine) CreateRole(roleJSONBytes []byte) ArboristOperation {
+	// Try to parse a role from the input.
+	var roleJSON *RoleJSON = &RoleJSON{}
+	err := json.Unmarshal(roleJSONBytes, roleJSON)
+	if err != nil {
+		msg := fmt.Sprintf("failed to parse role from JSON: %s", err)
+		return badRequest(msg)
+	}
+
+	// Once the role is parsed, try to load it into the engine.
+	err = engine.LoadRoleFromJSON(*roleJSON)
+	if err != nil {
+		msg := fmt.Sprintf("failed to load role: %s", err)
+		return badRequest(msg)
+	}
+
+	return ArboristOperation{
+		Success: true,
+		Status:  http.StatusNoContent,
+		JSON:    []byte{},
+	}
+}
+
+// WriteRole, given a role ID, writes out the information about the role.
+func (engine *AuthEngine) WriteRole(roleID string) ArboristOperation {
```

Comment:
nitpick.. why is it called Write? instead of GetRole for CRUD convention?

Reply:
yeah true, `Write` is really the opposite of what's meant here...fixed in 6d736a4